File: apps/ios/SmartAssist/Sources/Network/GatewayClient.swift
import Foundation

public actor GatewayClient {
    public var baseURL: URL?
    private var authToken: String?

    public init() {}

    public func setAuthToken(_ token: String) {
        self.authToken = token
    }

    public func sendRPC(method: String, params: [String: Any]? = nil) async throws -> [String: Any] {
        guard let baseURL = baseURL else {
            throw SmartAssistError.noGatewayURL
        }

        let url = baseURL.appendingPathComponent("rpc")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token = authToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let body: [String: Any] = [
            "jsonrpc": "2.0",
            "id": UUID().uuidString,
            "method": method,
            "params": params ?? [:]
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            throw SmartAssistError.rpcFailed
        }

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw SmartAssistError.invalidResponse
        }

        return json
    }

    public func pairDevice(deviceId: String, name: String) async throws -> String {
        let result = try await sendRPC(
            method: "device.pair",
            params: ["device_id": deviceId, "name": name, "device_type": "ios"]
        )
        guard let challenge = result["result"] as? [String: Any],
              let code = challenge["challenge_code"] as? String else {
            throw SmartAssistError.pairingFailed
        }
        return code
    }

    public func listDevices() async throws -> [Device] {
        let result = try await sendRPC(method: "device.pair.list")
        guard let response = result["result"] as? [String: Any],
              let devices = response["devices"] as? [[String: Any]] else {
            return []
        }
        return devices.compactMap { dict in
            guard let id = dict["id"] as? String,
                  let name = dict["name"] as? String,
                  let type = dict["device_type"] as? String else { return nil }
            var device = Device(id: id, name: name, deviceType: type)
            device.connected = dict["connected"] as? Bool ?? false
            return device
        }
    }
}

public enum SmartAssistError: Error {
    case noGatewayURL
    case rpcFailed
    case invalidResponse
    case pairingFailed
}

File: apps/android/app/src/main/java/com/smartassist/network/GatewayClient.kt
package com.smartassist.network

import io.ktor.client.HttpClient
import io.ktor.client.engine.android.Android
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.plugins.websocket.WebSockets
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.serialization.kotlinx.json.json
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonObject

class GatewayClient {
    var baseUrl: String = "http://localhost:18789"
    private var authToken: String? = null

    private val client = HttpClient(Android) {
        install(ContentNegotiation) {
            json(Json { ignoreUnknownKeys = true })
        }
        install(WebSockets)
    }

    fun setAuthToken(token: String) {
        authToken = token
    }

    suspend fun sendRPC(method: String, params: Map<String, String>? = null): JsonObject =
        withContext(Dispatchers.IO) {
            val body = JsonObject(
                mapOf(
                    "jsonrpc" to JsonPrimitive("2.0"),
                    "id" to JsonPrimitive(System.currentTimeMillis().toString()),
                    "method" to JsonPrimitive(method),
                    "params" to JsonObject(params ?: emptyMap())
                )
            )

            val response = client.post("$baseUrl/rpc") {
                contentType(ContentType.Application.Json)
                authToken?.let {
                    headers["Authorization"] = "Bearer $it"
                }
                setBody(body)
            }

            val text = response.bodyAsText()
            Json.parseToJsonElement(text).jsonObject
        }

    suspend fun pairDevice(deviceId: String, name: String): String =
        withContext(Dispatchers.IO) {
            val result = sendRPC(
                "device.pair",
                mapOf(
                    "device_id" to deviceId,
                    "name" to name,
                    "device_type" to "android"
                )
            )
            val response = result["result"]?.jsonObject
                ?: throw IllegalStateException("Pairing failed")
            response["challenge_code"]?.toString()?.trim('"')
                ?: throw IllegalStateException("No challenge code")
        }

    suspend fun listDevices(): List<com.smartassist.data.Device> =
        withContext(Dispatchers.IO) {
            val result = sendRPC("device.pair.list")
            val response = result["result"]?.jsonObject
            val devices = response?.get("devices")?.toString() ?: "[]"
            // Simplified parsing - in a real app use kotlinx.serialization
            emptyList()
        }
}
